ql/time/calendar.hpp

```diff
@@ -240,8 +246,16 @@ namespace QuantLib {
         return impl_->isBusinessDay(_d);
     }
 
+    inline bool Calendar::isStartOfMonth(const Date& d) const {
+        return d == startOfMonth(d);
+    }
+
+    inline Date Calendar::startOfMonth(const Date& d) const {
+        return adjust(Date::startOfMonth(d), Following);
+    }
+
     inline bool Calendar::isEndOfMonth(const Date& d) const {
-        return (d.month() != adjust(d+1).month());
+        return d == endOfMonth(d);
```

Comment:
Ok, no, my bad—the previous implementation returns true if a date is the last business day of its month but also if it's a holiday after the last business day of the month.  The check is used with this semantics while building schedules, which is why a few tests are failing.  We should revert and comment it properly.

I'm not sure what to do with `isStartOfMonth`.  I'm tempted to leave it as it is, even though it wouldn't be consistent with `isEndOfMonth`.  Any thoughts?

Reply:
Oh, I see what you mean. But the same holds for the start of month too? E.g. both 1 May 2024 and 2 May 2024 should return `true` for `isStartOfMonth()` if you observe International Workers Day?

Reply:
That's what I'm not sure about.  Ideally I'd return true only for the 2nd, but then it would be inconsistent with `isEndOfMonth`.  I guess we can keep them consistent and document it.

In other news, we can keep the simpler implementation of `isEndOfMonth` if we write it as

    return d >= endOfMonth(d);

i.e. with a `>=` instead of `==`.

Reply:
I definitely prefer them consistent. How about we add overload functions if we strictly want to test only for the first/last business day of the month?

Reply:
If one wants a strict test they can write `d == endOfMonth(d)`.